### PR TITLE
Ability statOffset

### PR DESCRIPTION
**Pull request by oorzkws**
Additive while statFactor is multiplicative. This allows for small increments to stat changes without multiplying the whole value. For example, a psycast with a base range of 30 could have an additional 5 range added per psylink level.

```
baseRange = 30
finalRange = 30 + (psychicSensitivity * 5)
```

[use case](https://steamcommunity.com/sharedfiles/filedetails/?id=2881380497)

### CODE REVIEW

**Comment by oorzkws** [comment]
I'm pretty sure this will be the last extension to Ability for a while, sorry for the repeated PRs.

File: Source/VFECore/Abilities/Ability.cs
Comment (oorzkws) [inline Source/VFECore/Abilities/Ability.cs:94]: I would like feedback on this. I'm not sure if this is the best way to check if a stat is expressed as a percentage relative to the base value.

Comment (legodude17) [reply Source/VFECore/Abilities/Ability.cs:94]: This seems unnecessarily confusing. Why can't you simply add the value of the stat to the current value?

File: Source/VFECore/Abilities/Ability.cs
Comment (legodude17) [inline Source/VFECore/Abilities/Ability.cs:100]:
```suggestion
            statOffsets.Aggregate(statFactors.Aggregate(current, CalculateStatFactorForPawn), CalculateStatOffsetForPawn);
```
Vanilla does offsets then factors, so to be consistent we should too.

File: Source/VFECore/Abilities/AbilityDef.cs
Comment (legodude17) [inline Source/VFECore/Abilities/AbilityDef.cs:44]:
```suggestion
        public List<StatModifier> castTimeStatOffsets = new List<StatModifier>();
```
Please be consistent with your spacing.

File: Source/VFECore/Abilities/AbilityDef.cs
Comment (legodude17) [inline Source/VFECore/Abilities/AbilityDef.cs:48]:
```suggestion
        public List<StatModifier> cooldownTimeStatOffsets = new List<StatModifier>();
```
Please be consistent with your spacing.

File: Source/VFECore/Abilities/AbilityDef.cs
Comment (legodude17) [inline Source/VFECore/Abilities/AbilityDef.cs:52]:
```suggestion
        public List<StatModifier> durationTimeStatOffsets = new List<StatModifier>();
```
Please be consistent with your spacing.

**Review by legodude17 — CHANGES_REQUESTED** [review]
Good idea, but there's a few issues to resolve first.

File: Source/VFECore/Abilities/Ability.cs
Comment (oorzkws) [reply Source/VFECore/Abilities/Ability.cs:100]: > This seems unnecessarily confusing. Why can't you simply add the value of the stat to the current value?
There are different kinds of stats - some are expressed as a percentage where we would only want to offset by the amount above/below the base level. Some are straight numbers, where we would want to use it as a true offset.

For example:
Psychic sensitivity - statOffset * (currentSensitivity - baseSensitivity)
Movespeed (c/s) - statOffset * moveSpeed

Comment (legodude17) [reply Source/VFECore/Abilities/Ability.cs:100]: I think you responded to the wrong comment.

File: Source/VFECore/Abilities/Ability.cs
Comment (oorzkws) [reply Source/VFECore/Abilities/Ability.cs:94]: > This seems unnecessarily confusing. Why can't you simply add the value of the stat to the current value?

There are different kinds of stats - some are expressed as a percentage where we would only want to offset by the amount above/below the base level. Some are straight numbers, where we would want to use it as a true offset.

For example:
Psychic sensitivity - statOffset * (currentSensitivity - baseSensitivity)
Movespeed (c/s) - statOffset * moveSpeed

Comment (oorzkws) [reply Source/VFECore/Abilities/Ability.cs:94]: There's one other thing I'm struggling with here - there are two "kinds" of offsets that could be desired. One expressed as a static value, and the other expressed as a percentage of the base value. To elaborate:

AbilityA has a range of 30.
I could desire to offset it by (5 * psychicSensitivity), which is how it currently works.
I could also desire to offset it by (5% of range * psychicSensitivity), which in this case would be (1.5 * psychicSensitivity)

I think the latter is more useful, but I'm not sure if this would be confusing behavior.

Comment (legodude17) [reply Source/VFECore/Abilities/Ability.cs:94]: > Psychic sensitivity - statOffset * (currentSensitivity - baseSensitivity)

Wouldn't this cause weird issues? Since if you are 150% psychic sensitivty the way that code would work is that it would add half the offset, which sounds not like what you want. I think it's a better bet to just add `offset * value`.

> There's one other thing I'm struggling with here - there are two "kinds" of offsets that could be desired. One expressed as a static value, and the other expressed as a percentage of the base value.

I think for anything more complex then static values we can simply say they have to use C# and override the corresponding calculator method.

Comment (oorzkws) [reply Source/VFECore/Abilities/Ability.cs:94]: > > Psychic sensitivity - statOffset * (currentSensitivity - baseSensitivity)
> 
> Wouldn't this cause weird issues? Since if you are 150% psychic sensitivty the way that code would work is that it would add half the offset, which sounds not like what you want. I think it's a better bet to just add `offset * value`.
> 
> > There's one other thing I'm struggling with here - there are two "kinds" of offsets that could be desired. One expressed as a static value, and the other expressed as a percentage of the base value.
> 
> I think for anything more complex then static values we can simply say they have to use C# and override the corresponding calculator method.

That would mean a stat at 100% would still affect the result, where it's instead desired (at least in my case) to scale the offset relative to how far the stat is above/below the base value.

For example:
Ability has a range of 30 and I add a sensitivity offset of 10. The offset indicates that the range will increase by 10 per 100% sensitivity above base.

Sensitivity 100%: 30 (base range)
Sensitivity 150%: 30 + 5
Sensitivity 200%: 30 + 10
Sensitivity 50%: 30 - 5

This is similar to how consciousness affects manipulation.

Comment (legodude17) [reply Source/VFECore/Abilities/Ability.cs:94]: Ok I see what you mean. I think I agree with you. Instead of making an array and checking `Contains` inline, can you make a `private static HashSet<ToStringStyle>` and check against that?

Comment (oorzkws) [reply Source/VFECore/Abilities/Ability.cs:94]: > Ok I see what you mean. I think I agree with you. Instead of making an array and checking `Contains` inline, can you make a `private static HashSet<ToStringStyle>` and check against that?

Sure, something to this effect? I moved the ternary statement to an if block to make the syntax clearer.
```c#
        private static readonly HashSet<ToStringStyle> percentageBasedStyles = new HashSet<ToStringStyle>
        {
            ToStringStyle.PercentZero,
            ToStringStyle.PercentOne,
            ToStringStyle.PercentTwo
        };
        
        private float CalculateStatOffsetForPawn(float current, StatModifier statOffset)
        {
            if (percentageBasedStyles.Contains(statOffset.stat.toStringStyle))
            {
                return current + (this.pawn.GetStatValue(statOffset.stat) - statOffset.stat.defaultBaseValue) *
                    statOffset.value;
            }

            return current + this.pawn.GetStatValue(statOffset.stat) * statOffset.value;
        }
```

Comment (legodude17) [reply Source/VFECore/Abilities/Ability.cs:94]: Yeah exactly, though I don't mind the ternary.

Comment (oorzkws) [reply Source/VFECore/Abilities/Ability.cs:94]: Great. If I'm expecting modders to override this in harmony (for the percentage offset, for example) should the method be public?

Comment (legodude17) [reply Source/VFECore/Abilities/Ability.cs]: No, it's an internal method, so leave it private. People can use the name if they need to patch it.

**Comment by oorzkws** [comment]
Thanks for the thorough review and input!

**Review by legodude17 — APPROVED** [review]
Looks good!

**Comment by oorzkws** [comment]
I have rebased this on current main (99c955e)

**Comment by legodude17** [comment]
Waiting on approval from @erdelf.

**Comment by erdelf** [comment]
While I give it a proper review, kill those meta files, and definitely get the change to the references and included files out of there.

**Review by erdelf — CHANGES_REQUESTED** [review]
Remove the .idea meta files and the change to the references.
Change your new methods to virtual and set the accessor accordingly.

**Comment by oorzkws** [comment]
> Remove the .idea meta files and the change to the references. 

Yeah that was my mistake in rebasing. Resolved.

> Change your new methods to virtual and set the accessor accordingly.

Done.

**Review by erdelf — APPROVED** [review]

**Comment by legodude17** [comment]
Thank you!

**Comment by oorzkws** [comment]
> Thank you!

Thanks for the thorough feedback, it's always appreciated to have reviews beyond "looks good to me". You've seen there's a PR open on VPE, yeah? No hurry, just seems few have notifications on there.

**Pull request merged**
